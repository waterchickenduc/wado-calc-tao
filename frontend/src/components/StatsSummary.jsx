import React from "react";

function formatValue(value) {
  return `${value.toFixed(2)}%`;
}

function StatBlock({ title, stats, color = "text-white", subTitle = null }) {
  const keys = Object.keys(stats);
  if (keys.length === 0) return null;

  return (
    <div className="mt-4">
      <h4 className={`font-bold ${color}`}>{title}</h4>
      {subTitle && <div className="text-yellow-500 text-xs mb-1">{subTitle}</div>}
      {keys.map((key) => (
        <div key={key} className="flex justify-between text-white">
          <span>{key}</span>
          <span>{formatValue(stats[key])}</span>
        </div>
      ))}
    </div>
  );
}

function groupStatsByAura(runes) {
  const grouped = {};
  for (const rune of runes) {
    if (!rune.aura || !rune.auraChance) continue;

    const auraKey = rune.aura;
    if (!grouped[auraKey]) {
      grouped[auraKey] = {
        triggerChance: 0,
        stats: {},
      };
    }

    grouped[auraKey].triggerChance += rune.auraChance;

    const auraStats = rune?.auraStats || [];
    for (const { Stat, Value } of auraStats) {
      grouped[auraKey].stats[Stat] = (grouped[auraKey].stats[Stat] || 0) + Value;
    }
  }

  return grouped;
}

export default function StatsSummary({ totalStats, runeStats, auraStats, classStats, runes = [] }) {
  const groupedAuraStats = groupStatsByAura(runes);

  return (
    <div className="bg-zinc-900 border border-zinc-700 p-4 rounded w-full md:max-w-xs text-sm">
      <h3 className="text-white font-semibold mb-4 text-lg">ðŸ“Š Stats Summary</h3>

      <StatBlock title="Total" stats={totalStats} color="text-pink-400" />
      <StatBlock title="Total by Runes" stats={runeStats} color="text-green-400" />

      {/* ðŸŒŸ Per-Aura Breakdown */}
      {Object.keys(groupedAuraStats).length > 0 && (
        <>
          <h4 className="text-yellow-400 font-bold mt-4">Total by Auras</h4>
          {Object.entries(groupedAuraStats).map(([auraName, data]) => (
            <div key={auraName} className="mt-3">
              <div className="text-yellow-300 font-medium">{auraName}</div>
              <div className="text-yellow-500 text-xs mb-1">
                Trigger Chance: {formatValue(data.triggerChance)}
              </div>
              {Object.entries(data.stats).map(([stat, value]) => (
                <div key={stat} className="flex justify-between text-yellow-200">
                  <span>{stat}</span>
                  <span>{formatValue(value)}</span>
                </div>
              ))}
            </div>
          ))}
        </>
      )}

      <StatBlock title="Total by Classes" stats={classStats} color="text-blue-400" />
    </div>
  );
}
