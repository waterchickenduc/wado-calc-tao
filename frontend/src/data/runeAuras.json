[
  {
    "name": "Concentration",
    "trigger": 10,
    "stats": [
      {
        "Stat": "Cr. Rate",
        "Value": 25
      }
    ]
  },
  {
    "name": "Conviction",
    "trigger": 10,
    "stats": [
      {
        "Stat": "P. Atk",
        "Value": 15
      }
    ]
  },
  {
    "name": "Dream",
    "trigger": 10,
    "stats": [
      {
        "Stat": "HP Recovery",
        "Value": 15
      }
    ]
  },
  {
    "name": "Meditation",
    "trigger": 10,
    "stats": [
      {
        "Stat": "M. Atk",
        "Value": 15
      }
    ]
  },
  {
    "name": "Prayer",
    "trigger": 10,
    "stats": [
      {
        "Stat": "P. Def",
        "Value": 15
      },
      {
        "Stat": "M. Def",
        "Value": 15
      },
      {
        "Stat": "Movementspeed",
        "Value": 10
      }
    ]
  },
  {
    "name": "Protect",
    "trigger": 10,
    "stats": [
      {
        "Stat": "P. Def",
        "Value": 50
      },
      {
        "Stat": "M. Def",
        "Value": 50
      }
    ]
  },
  {
    "name": "Sanctuary",
    "trigger": 10,
    "stats": [
      {
        "Stat": "P. Atk",
        "Value": 10
      },
      {
        "Stat": "P. Def",
        "Value": 10
      },
      {
        "Stat": "M. Atk",
        "Value": 10
      },
      {
        "Stat": "M. Def",
        "Value": 10
      }
    ]
  },
  {
    "name": "Thorns",
    "trigger": 10,
    "stats": [
      {
        "Stat": "P. Atk",
        "Value": 35
      },
      {
        "Stat": "M. Atk",
        "Value": 35
      }
    ]
  },
  {
    "name": "Salvation",
    "trigger": 10,
    "stats": [
      {
        "Stat": "Increase STR",
        "Value": 5
      },
      {
        "Stat": "Increase AGI",
        "Value": 5
      },
      {
        "Stat": "Increase INT",
        "Value": 5
      }
    ]
  },
  {
    "name": "Might",
    "trigger": 10,
    "stats": [
      {
        "Stat": "Skill CD",
        "Value": 5
      }
    ]
  }
]