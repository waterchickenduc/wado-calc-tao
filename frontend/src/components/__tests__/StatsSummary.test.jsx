import React from "react";
import { render, screen } from "@testing-library/react";
import StatsSummary from "../StatsSummary";
import runeAuras from "../../data/runeAuras.json";

const mockRunes = [
  {
    name: "Petra",
    aura: "Thorns (3%)",
    auraChance: 3,
    stats: [
      { Stat: "Ele. Def", Value: 7 },
      { Stat: "Increase INT", Value: 0.25 },
    ]
  },
  {
    name: "Petra",
    aura: "Thorns (3%)",
    auraChance: 3,
    stats: [
      { Stat: "Ele. Def", Value: 7 },
      { Stat: "Increase INT", Value: 0.25 },
    ]
  },
  {
    name: "Draumr",
    aura: "Dream (2%)",
    auraChance: 2,
    stats: [{ Stat: "Cr. Rate", Value: 3 }]
  },
  {
    name: "Darkstone",
    aura: "Corrupt (5%)",
    auraChance: 5,
    stats: []
  }
];

const mockClassStats = {
  "Ele. Def": 5,
  "P. Atk": 10
};

describe("StatsSummary", () => {
  test("does not render aura section if no runes have aura", () => {
    render(<StatsSummary runeStats={{}} classStats={{}} runes={[]} />);
    expect(screen.queryByText("Total by Auras")).not.toBeInTheDocument();
  });

  test("renders Total, Runes, Auras, Classes correctly with multiple aura types", () => {
    render(
      <StatsSummary runeStats={{}} classStats={mockClassStats} runes={mockRunes} />
    );

    // âœ… Section checks
    expect(screen.getByText("Total")).toBeInTheDocument();
    expect(screen.getByText("Total by Runes")).toBeInTheDocument();
    expect(screen.getByText("Total by Auras")).toBeInTheDocument();
    expect(screen.getByText("Total by Classes")).toBeInTheDocument();

    // âœ… Aura blocks
    expect(screen.getByText("Thorns")).toBeInTheDocument();
    expect(screen.getByText("Dream")).toBeInTheDocument();
    expect(screen.getByText("Corrupt")).toBeInTheDocument();

    // ðŸ§ª Trigger chance & stats
    expect(screen.getByText(/Trigger Chance:\s*6.00/)).toBeInTheDocument();
    expect(screen.getByText(/Trigger Chance:\s*2.00/)).toBeInTheDocument();
    expect(screen.getByText(/Trigger Chance:\s*5.00/)).toBeInTheDocument();

    // ðŸ“Œ Safe numeric value assertions
    expect(screen.getAllByText("15.00%").length).toBeGreaterThanOrEqual(1); // HP Recovery
    expect(screen.getByText("70.00%")).toBeInTheDocument(); // M. Atk or P. Atk from Thorns
    expect(screen.getByText("10.00%")).toBeInTheDocument(); // Class Ele. Def or Corrupt Ele. Def
  });

  test("snapshot of full render", () => {
    const { container } = render(
      <StatsSummary runeStats={{}} classStats={mockClassStats} runes={mockRunes} />
    );
    expect(container).toMatchSnapshot();
  });
});
