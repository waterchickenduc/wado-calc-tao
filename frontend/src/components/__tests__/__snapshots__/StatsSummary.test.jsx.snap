// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StatsSummary > matches snapshot of full render 1`] = `
<div>
  <div
    class="bg-zinc-900 border border-zinc-700 p-4 rounded w-full md:max-w-xs text-sm"
  >
    <h3
      class="text-white font-semibold mb-4 text-lg"
    >
      ðŸ“Š Stats Summary
    </h3>
    <div
      class="mt-4"
    >
      <h4
        class="font-bold text-pink-400"
      >
        Total
      </h4>
      <div
        class="flex justify-between text-white"
      >
        <span>
          P. Atk
        </span>
        <span>
          10.00%
        </span>
      </div>
      <div
        class="flex justify-between text-white"
      >
        <span>
          Cr. Rate
        </span>
        <span>
          3.00%
        </span>
      </div>
      <div
        class="flex justify-between text-white"
      >
        <span>
          M. Def
        </span>
        <span>
          10.00%
        </span>
      </div>
      <div
        class="flex justify-between text-white"
      >
        <span>
          Increase INT
        </span>
        <span>
          0.25%
        </span>
      </div>
    </div>
    <div
      class="mt-4"
    >
      <h4
        class="font-bold text-green-400"
      >
        Total by Runes
      </h4>
      <div
        class="flex justify-between text-white"
      >
        <span>
          Cr. Rate
        </span>
        <span>
          3.00%
        </span>
      </div>
    </div>
    <h4
      class="text-yellow-400 font-bold mt-4"
    >
      Total by Auras
    </h4>
    <div
      class="mt-3"
    >
      <div
        class="text-yellow-300 font-medium"
      >
        Sanctuary
      </div>
      <div
        class="text-yellow-500 text-xs mb-1"
      >
        Trigger Chance: 
        5.00%
      </div>
      <div
        class="flex justify-between text-yellow-200"
      >
        <span>
          P. Atk
        </span>
        <span>
          10.00%
        </span>
      </div>
      <div
        class="flex justify-between text-yellow-200"
      >
        <span>
          M. Def
        </span>
        <span>
          10.00%
        </span>
      </div>
    </div>
    <div
      class="mt-4"
    >
      <h4
        class="font-bold text-blue-400"
      >
        Total by Classes
      </h4>
      <div
        class="flex justify-between text-white"
      >
        <span>
          Increase INT
        </span>
        <span>
          0.25%
        </span>
      </div>
    </div>
  </div>
</div>
`;
